refactor 'Contribute Code' (#130)
pingcap/tidb-dev-guide: src/contribute-to-tidb/contribute-code.md

# Contribute Code

TiDB is maintained, improved, and extended by code contributions. We welcome code contributions to TiDB. TiDB uses a workflow based on [pull requests](https://docs.github.com/en/github/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests).

## Before contributing

> Contributing to TiDB does *not* start with opening a pull request. We expect contributors to reach out to us first to discuss the overall approach together. Without consensus with the TiDB committers, contributions might require substantial rework or will not be reviewed. So please [create a GitHub issue](open-an-issue.md) or create a topic on the [internal.tidb.io](https://internals.tidb.io) and reach consensus.

Reviewer: ```suggestion
Contributing to TiDB does *not* start with opening a pull request. We expect contributors to reach out to us first to discuss the overall approach together. Without consensus with the TiDB committers, contributions might require substantial rework or will not be reviewed. So please [create a GitHub issue](open-an-issue.md) or discuss under an existing issue or create a topic on the [internal.tidb.io](https://internals.tidb.io) and reach consensus.
```

# Contribute Code

TiDB is maintained, improved, and extended by code contributions. We welcome code contributions to TiDB. TiDB uses a workflow based on [pull requests](https://docs.github.com/en/github/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests).

## Before contributing

Contributing to TiDB does *not* start with opening a pull request. We expect contributors to reach out to us first to discuss the overall approach together. Without consensus with the TiDB committers, contributions might require substantial rework or will not be reviewed. So please [create a GitHub issue](open-an-issue.md), discuss under an existing issue or create a topic on the [internal.tidb.io](https://internals.tidb.io) and reach consensus.

> For newcomers, you can check the [starter issues](https://github.com/pingcap/tidb/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22), which are annotated with a "good first issue" label. These are issues suitable for new contributors to work with and won't take long to fix. But because the label is typically added at triage time it can turn out to be inaccurate, so do feel free to leave a comment if you think the classification no longer applies.

Author: Nice. Better.

> To get your change merged you need to sign the [CLA](https://cla-assistant.io/pingcap/tidb?pullRequest=5567) to grant PingCAP ownership of your code.

Author: Actually I prefer to tell the user the CLA is a must. We just tell them it's really needed. If they really want to know detail, they should read the full CLA document.

## Contributing process

After a consensus is reached in issues, it's time to start the code contributing process:

  1. Assign the issue to yourself via [\assign](https://prow.tidb.io/command-help?repo=pingcap%2Ftidb#assign). This lets other contributors know you are working on the issue so they won't make duplicate efforts.
  2. Follow the [GitHub workflow](https://guides.github.com/introduction/flow/), commit code changes in your own git repository branch and open a pull request for code review.
  3. Make sure the continuous integration checks on your pull request are green (i.e. successful).
  4. Review and address [comments on your pull request](https://devguide.python.org/pullrequest/). If your pull request becomes outdated with the target branch, you need to [rebase your pull request](https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request#perform-a-rebase) to keep it up to date.
  5. When your pull request gets enough approvals(the default number is 2) and all other requirements are met, it will be merged.

>   6. Handle regressions introduced by your change. Although committers bear the main responsibility to fix regressions, it's quite nice for you to handle it(reverting the change or sending fixes).

Reviewer 2: Several comments:

1. I don't think we should link to Python Dev Guide here. If we want any of its content, write our own.
2. Because of we use "squash and merge" we don't have to rebase on master but often just merge master to catch up change.

>   6. Handle regressions introduced by your change. Although committers bear the main responsibility to fix regressions, it's quite nice for you to handle it(reverting the change or sending fixes).

Author: 1. It's my mistake. Seems I paste a wrong link, it should be a link to a official GitHub guide.
2. Git practise could vary between people. merge and rebase should both work here.

Clear and kind communication is key to this process.

## Writing tests

One important thing when you make code contributions to TiDB is tests. Tests should be always considered as a part of your change. Any code changes that cause semantic changes or new function additions to TiDB should have corresponding test cases. And of course you can not break any existing test cases if they are still valid. It's recommended to [run tests](../get-started/write-and-run-unit-tests.md) on your local environment first to find obvious problems and fix them before opening the pull request.

It's also highly appreciated if your pull request only contains test cases to increase test coverage of TiDB. Supplement test cases for existing modules is a good and easy way to become acquainted with existing code.

## Making good pull requests

When creating a pull request for submission, there are several things that you should consider to help ensure that your pull request is accepted:

>   - Does the contribution alter the behavior of features or components in a way that it may break previous users’ programs and setups? If yes, there needs to be a discussion and agreement that this change is desirable.
>   - Does the contribution conceptually fit well into TiDB? Is it too much of a special case such that it makes things more complicated for the common case, or bloats the abstractions / APIs?
>   - Does the contribution make a big impact on TiDB's build time?
>   - Does your contribution affect any documentation? If yes, you should add/change proper documentation. 
>   - If there are any new dependencies, are they under active maintainences? What are their licenses?

Reviewer 2: ```suggestion
* Does the contribution alter the behavior of features or components in a way that it may break previous users' programs and setups? If yes, there needs to be a discussion and agreement that this change is desirable.
* Does the contribution conceptually fit well into TiDB? Is it too much of a special case such that it makes things more complicated for the common case, or bloats the abstractions/APIs?
* Does the contribution make a big impact on TiDB's build time?
* Does your contribution affect any documentation? If yes, you should add/change proper documentation. 
* If there are any new dependencies, are they under active maintenances? What are their licenses?
```

## Making good commits

Each feature or bugfix should be addressed by a single pull request, and for each pull request there may be several commits. In particular:

>   - Do *not* fix more than one issue in the same commit (except, of course, if one code change fixes all of them).
>   - Do *not* do cosmetic changes to unrelated code in the same commit as some feature/bugfix.

Reviewer 2: ```suggestion
* Do *not* fix more than one issues in the same commit (except, of course, if one code change fixes all of them).
* Do *not* do cosmetic changes to unrelated code in the same commit as some feature/bugfix.
```

## Waiting for review

> To begin with, please be patient! There are many more people submitting pull requests than there are people capable of reviewing your pull request. Getting your pull request reviewed requires a reviewer to have the spare time and motivation to look at your pull request. If your pull request has not received any notice from reviewers (i.e., no comment made) for some time, you can ping the reviewers and asignees, or take it to [internal.tidb.io](https://internals.tidb.io) for more attention.

Reviewer 2: ```suggestion
To begin with, please be patient! There are many more people submitting pull requests than there are people capable of reviewing your pull request. Getting your pull request reviewed requires a reviewer to have the spare time and motivation to look at your pull request. If your pull request has not received any notice from reviewers (i.e., no comment made) for some time, you can ping the reviewers and assignees, or take it to [internal.tidb.io](https://internals.tidb.io) for more attention.
```

> When someone does manage to find the time to look at your pull request they will most likely make comments about how it can be improved (don’t worry, even committers/maintainers have their pull requests sent back to them for changes). It is then expected that you update your pull request to address these comments, and the review process will thus iterate until a satisfactory solution has emerged.

Reviewer 2: ```suggestion
When someone does manage to find the time to look at your pull request, they will most likely make comments about how it can be improved (don't worry, even committers/maintainers have their pull requests sent back to them for changes). It is then expected that you update your pull request to address these comments, and the review process will thus iterate until a satisfactory solution has emerged.
```
